File: src/graph/flight_discount.py
import heapq


n, m = map(int, input().split())

adjList = [[] for _ in range(n + 1)]
revAdjList = [[] for _ in range(n + 1)]

for i in range(m):
    a, b, c = map(int, input().split())
    adjList[a].append((b, c))
    revAdjList[b].append((a, c))


def dijskra(i, j, adjList):
    d = [float("inf") for _ in range(n + 1)]
    d[i] = 0

    visited = [False] * (n + 1)

    queue = []
    heapq.heappush(queue, (0, 0, i))

    while queue:
        dist, _max, node = heapq.heappop(queue)
        # print("dist", d)
        # print("pop", dist, _max, node)

        if visited[node]:
            continue

        visited[node] = True

        if node == j:
            return d[j]

        for adj, wt in adjList[node]:
            if wt > _max:
                _new_dist = dist + (_max - _max // 2) + wt // 2
            else:
                _new_dist = dist + wt
            # print("adj", adj, node, _new_dist, d[adj])
            if _new_dist <= d[adj]:
                d[adj] = _new_dist
                # print("push", (d[adj], max(_max, wt), adj))
                heapq.heappush(queue, (d[adj], max(_max, wt), adj))


# print(adjList)

print(min(dijskra(1, n, adjList), dijskra(n, 1, revAdjList)))

File: src/graph/flight_discount.rs
use std::cmp::Ordering;
use std::collections::BinaryHeap;

#[derive(Copy, Clone, Eq, PartialEq)]
struct Node {
    dist: u64,
    max_weight: u64,
    id: usize,
}

impl Ord for Node {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .dist
            .cmp(&self.dist)
            .then_with(|| other.max_weight.cmp(&self.max_weight))
    }
}

impl PartialOrd for Node {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(start: usize, end: usize, adj_list: &[Vec<(usize, u64)>]) -> u64 {
    let n = adj_list.len();
    let mut distances = vec![std::u64::MAX; n];
    distances[start] = 0;

    let mut visited = vec![false; n];
    let mut queue = BinaryHeap::new();

    queue.push(Node {
        dist: 0,
        max_weight: 0,
        id: start,
    });

    while let Some(Node {
        dist,
        max_weight,
        id,
    }) = queue.pop()
    {
        if visited[id] {
            continue;
        }

        visited[id] = true;

        if id == end {
            return distances[end];
        }

        for &(adj, weight) in &adj_list[id] {
            let new_dist = if weight > max_weight {
                dist + (max_weight - max_weight / 2) + weight / 2
            } else {
                dist + weight
            };

            if new_dist <= distances[adj] {
                distances[adj] = new_dist;
                queue.push(Node {
                    dist: new_dist,
                    max_weight: max_weight.max(weight),
                    id: adj,
                });
            }
        }
    }
    std::u64::MAX
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let nm: Vec<usize> = input
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let n = nm[0];
    let m = nm[1];

    let mut adj_list = vec![Vec::new(); n + 1];
    let mut rev_adj_list = vec![Vec::new(); n + 1];

    for _ in 0..m {
        input.clear();
        std::io::stdin().read_line(&mut input).unwrap();
        let abc: Vec<usize> = input
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        let a = abc[0];
        let b = abc[1];
        let c = abc[2];
        adj_list[a].push((b, c as u64));
        rev_adj_list[b].push((a, c as u64));
    }

    let result = dijkstra(1, n, &adj_list).min(dijkstra(n, 1, &rev_adj_list));
    println!("{}", result);
}
